File: backend/app/schemas/admin_ai.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class AiModelApiBase(BaseModel):
    name: str = Field(..., description="展示名称")
    provider: str = Field(..., description="dashscope_openai / ark_responses")
    provider_brand: Optional[str] = Field(None, description="OpenAI / DeepSeek / AzureOpenAI 等")
    model_name: str = Field(..., description="模型名/版本")
    endpoint: str = Field(..., description="接口地址或 base url")
    api_key: str = Field(..., description="API Key")
    api_header: Optional[str] = Field(None, description="额外 Header（JSON 串）")
    api_version: Optional[str] = Field(None, description="API 版本号（Azure/OpenAI）")
    timeout_seconds: int = Field(30, ge=1, le=600)
    quota_per_hour: int = Field(0, ge=0)
    temperature: Optional[float] = Field(None, ge=0, le=2)
    max_output_tokens: Optional[int] = Field(None, ge=32, le=32768)
    enabled: bool = True
    is_default: bool = False


class AiModelApiCreate(AiModelApiBase):
    pass


class AiModelApiUpdate(BaseModel):
    name: Optional[str] = None
    provider: Optional[str] = None
    provider_brand: Optional[str] = None
    model_name: Optional[str] = None
    endpoint: Optional[str] = None
    api_key: Optional[str] = None
    api_header: Optional[str] = None
    api_version: Optional[str] = None
    timeout_seconds: Optional[int] = Field(default=None, ge=1, le=600)
    quota_per_hour: Optional[int] = Field(default=None, ge=0)
    temperature: Optional[float] = Field(default=None, ge=0, le=2)
    max_output_tokens: Optional[int] = Field(default=None, ge=32, le=32768)
    enabled: Optional[bool] = None
    is_default: Optional[bool] = None


class AiModelApiOut(BaseModel):
    id: int
    name: str
    provider: str
    provider_brand: Optional[str] = None
    model_name: str
    endpoint: str
    api_header: Optional[str] = None
    api_version: Optional[str] = None
    timeout_seconds: int
    quota_per_hour: int
    temperature: Optional[float] = None
    max_output_tokens: Optional[int] = None
    enabled: bool
    is_default: bool
    created_at: datetime
    updated_at: datetime


class AiKnowledgeBaseCreate(BaseModel):
    name: str
    slug: Optional[str] = None
    description: Optional[str] = None
    feature: Optional[str] = None
    is_default: Optional[bool] = False


class AiKnowledgeBaseUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    feature: Optional[str] = None
    is_default: Optional[bool] = None


class AiKnowledgeBaseOut(BaseModel):
    id: int
    slug: str
    name: str
    description: Optional[str] = None
    owner_type: str
    owner_user_id: Optional[int] = None
    course_id: Optional[int] = None
    feature: Optional[str] = None
    is_default: bool = False
    document_count: int = 0
    chunk_count: int = 0
    created_at: datetime
    updated_at: Optional[datetime] = None


class AiModelApiTestRequest(BaseModel):
    api_id: Optional[int] = None
    provider: str
    model_name: str
    endpoint: str
    api_key: Optional[str] = None
    timeout_seconds: int = 30
    prompt: Optional[str] = None


class AiModelApiTestResponse(BaseModel):
    ok: bool
    message: str
    output: Optional[str] = None


class AiKbSubjectCreate(BaseModel):
    name: str
    stage: Optional[str] = None


class AiKbSubjectOut(BaseModel):
    id: int
    name: str
    stage: Optional[str] = None
    enabled: bool = True
    created_at: datetime


class AiKbDocumentOut(BaseModel):
    id: int
    subject_id: int
    knowledge_base_id: Optional[int] = None
    title: str
    original_filename: str
    url: str
    file_ext: str
    file_size: int
    enabled: bool = True
    created_at: datetime
    updated_at: Optional[datetime] = None
    chunk_count: Optional[int] = None


class AiModelKbUpdateRequest(BaseModel):
    kb_document_ids: List[int] = Field(default_factory=list)


class AiCustomModelKbDocOut(BaseModel):
    id: int
    title: str
    subject_id: int
    subject_name: Optional[str] = None


class AiCustomModelCreate(BaseModel):
    name: str
    remark: Optional[str] = None
    enabled: bool = True
    base_model_api_id: int
    primary_subject_id: Optional[int] = None
    kb_document_ids: List[int] = Field(default_factory=list)


class AiCustomModelUpdate(BaseModel):
    name: Optional[str] = None
    remark: Optional[str] = None
    enabled: Optional[bool] = None
    base_model_api_id: Optional[int] = None
    primary_subject_id: Optional[int] = None
    kb_document_ids: Optional[List[int]] = None


class AiCustomModelOut(BaseModel):
    id: int
    name: str
    remark: Optional[str] = None
    enabled: bool
    base_model_api_id: int
    base_model_name: Optional[str] = None
    primary_subject_id: Optional[int] = None
    primary_subject_name: Optional[str] = None
    kb_documents: List[AiCustomModelKbDocOut] = Field(default_factory=list)
    created_at: datetime
    updated_at: Optional[datetime] = None


class AiFeatureBindingOut(BaseModel):
    feature: str
    custom_model_ids: List[int] = Field(default_factory=list)


class AiFeatureBindingUpdate(BaseModel):
    custom_model_ids: List[int] = Field(default_factory=list)


class AiCustomerServiceSettingsOut(BaseModel):
    welcome_str: str = ""
    recommend_questions: List[str] = Field(default_factory=list)
    search_placeholder: str = ""
    system_prompt_template: Optional[str] = None


class AiCustomerServiceSettingsUpdate(BaseModel):
    welcome_str: Optional[str] = None
    recommend_questions: Optional[List[str]] = None
    search_placeholder: Optional[str] = None
    system_prompt_template: Optional[str] = None


class PublicAiModelOut(BaseModel):
    id: int
    name: str
    provider: str
    model_name: str
    is_default: bool
    subjects: List[str] = Field(default_factory=list)


class AiWorkflowAppOut(BaseModel):
    code: str
    type: str
    name: str
    status: str
    knowledge_base_id: Optional[int] = None
    model_api_id: Optional[int] = None
    owner_user_id: Optional[int] = None
    course_id: Optional[int] = None
    settings: Dict[str, Any] = Field(default_factory=dict)
    updated_at: Optional[datetime] = None


class AiWorkflowAppCreate(BaseModel):
    code: str
    type: str
    name: str
    status: Optional[str] = "enabled"
    knowledge_base_id: Optional[int] = None
    model_api_id: Optional[int] = None
    owner_user_id: Optional[int] = None
    course_id: Optional[int] = None
    settings: Optional[Dict[str, Any]] = None


class AiWorkflowAppUpdate(BaseModel):
    name: Optional[str] = None
    knowledge_base_id: Optional[int] = None
    model_api_id: Optional[int] = None
    status: Optional[str] = None
    settings: Optional[Dict[str, Any]] = None


class StudentCourseAiSelectRequest(BaseModel):
    course_id: int
    model_api_id: Optional[int] = None


class StudentCourseAiSelectOut(BaseModel):
    course_id: int
    model_api_id: Optional[int] = None

File: frontend/src/api/adminAi.ts
import axios from 'axios'

export type AiProvider = 'dashscope_openai' | 'ark_responses'

export type AiFeatureKey = 'customer_service' | 'lesson_plan' | 'course_assistant'

export interface AiModelApiItem {
  id: number
  name: string
  provider: AiProvider
  provider_brand?: string
  model_name: string
  endpoint: string
  api_header?: string | null
  api_version?: string | null
  timeout_seconds: number
  quota_per_hour: number
  temperature?: number | null
  max_output_tokens?: number | null
  enabled: boolean
  is_default: boolean
  created_at: string
  updated_at: string
}

export interface AiModelApiCreate {
  name: string
  provider: AiProvider
  provider_brand?: string
  model_name: string
  endpoint: string
  api_key: string
  api_header?: string | null
  api_version?: string | null
  timeout_seconds: number
  quota_per_hour: number
  temperature?: number | null
  max_output_tokens?: number | null
  enabled: boolean
  is_default: boolean
}

export type AiModelApiUpdate = Partial<AiModelApiCreate>

export interface AiModelApiTestRequest {
  api_id?: number
  provider: AiProvider
  model_name: string
  endpoint: string
  api_key: string
  timeout_seconds: number
  prompt?: string
}

export interface AiModelApiTestResponse {
  ok: boolean
  message: string
  output?: string
}

export interface AiKbSubject {
  id: number
  name: string
  stage?: string | null
  enabled?: boolean
  created_at: string
}

export interface AiKbDocument {
  id: number
  subject_id: number
  knowledge_base_id?: number
  title: string
  original_filename: string
  url: string
  file_ext: string
  file_size: number
  enabled?: boolean
  created_at: string
  updated_at?: string
  chunk_count?: number
}

export interface AiCustomModelItem {
  id: number
  name: string
  remark?: string | null
  enabled: boolean
  base_model_api_id: number
  base_model_name?: string
  primary_subject_id?: number | null
  primary_subject_name?: string | null
  kb_documents?: Array<{ id: number; title: string; subject_id: number; subject_name?: string }>
  created_at: string
  updated_at?: string
}

export interface AiCustomModelCreate {
  name: string
  remark?: string
  enabled: boolean
  base_model_api_id: number
  primary_subject_id?: number | null
  kb_document_ids: number[]
}

export type AiCustomModelUpdate = Partial<AiCustomModelCreate>

export interface AiFeatureBindingOut {
  feature: AiFeatureKey
  custom_model_ids: number[]
}

export interface AiCustomerServiceSettings {
  welcome_str: string
  recommend_questions: string[]
  search_placeholder: string
  system_prompt_template?: string | null
}

export interface AiKnowledgeBaseItem {
  id: number
  slug: string
  name: string
  description?: string | null
  owner_type: string
  owner_user_id?: number | null
  course_id?: number | null
  feature?: string | null
  is_default: boolean
  document_count: number
  chunk_count: number
  created_at: string
  updated_at?: string | null
}

export interface AiKnowledgeBaseCreate {
  name: string
  slug?: string
  description?: string
  feature?: string
  is_default?: boolean
}

export interface AiWorkflowAppItem {
  code: string
  type: string
  name: string
  status: string
  knowledge_base_id?: number | null
  model_api_id?: number | null
  owner_user_id?: number | null
  course_id?: number | null
  settings: Record<string, any>
  updated_at?: string | null
}

export interface AiWorkflowAppUpdate {
  name?: string
  knowledge_base_id?: number | null
  model_api_id?: number | null
  status?: string
  settings?: Record<string, any>
}

// Axios 全局已设置 baseURL=/api（main.ts），这里保持相对路径，避免 /api/api 重复
const BASE_URL = '/admin/ai'

const authHeaders = () => {
  const token = localStorage.getItem('token')
  return token ? { Authorization: `Bearer ${token}` } : {}
}

export const adminAiApi = {
  async listModelApis() {
    const res = await axios.get<AiModelApiItem[]>(`${BASE_URL}/model-apis`, { headers: authHeaders() })
    return res.data
  },
  async createModelApi(payload: AiModelApiCreate) {
    const res = await axios.post<AiModelApiItem>(`${BASE_URL}/model-apis`, payload, { headers: authHeaders() })
    return res.data
  },
  async updateModelApi(id: number, payload: AiModelApiUpdate) {
    const res = await axios.put<AiModelApiItem>(`${BASE_URL}/model-apis/${id}`, payload, { headers: authHeaders() })
    return res.data
  },
  async deleteModelApi(id: number) {
    await axios.delete(`${BASE_URL}/model-apis/${id}`, { headers: authHeaders() })
  },
  async testModelApi(payload: AiModelApiTestRequest) {
    const res = await axios.post<AiModelApiTestResponse>(`${BASE_URL}/model-apis/test`, payload, { headers: authHeaders() })
    return res.data
  },
  async updateModelKbLinks(modelApiId: number, kbDocumentIds: number[]) {
    const res = await axios.put(`${BASE_URL}/model-apis/${modelApiId}/kb`, { kb_document_ids: kbDocumentIds }, { headers: authHeaders() })
    return res.data
  },

  async listSubjects() {
    const res = await axios.get<AiKbSubject[]>(`${BASE_URL}/kb/subjects`, { headers: authHeaders() })
    return res.data
  },
  async createSubject(name: string) {
    const res = await axios.post<AiKbSubject>(`${BASE_URL}/kb/subjects`, { name }, { headers: authHeaders() })
    return res.data
  },
  async listDocuments(subjectId: number) {
    const res = await axios.get<AiKbDocument[]>(`${BASE_URL}/kb/documents`, {
      headers: authHeaders(),
      params: { subject_id: subjectId }
    })
    return res.data
  },
  async uploadDocument(subjectId: number, title: string, file: File) {
    const form = new FormData()
    form.append('subject_id', String(subjectId))
    form.append('title', title)
    form.append('file', file)
    const res = await axios.post<AiKbDocument>(`${BASE_URL}/kb/upload`, form, {
      headers: { ...authHeaders() }
    })
    return res.data
  },
  async deleteDocument(docId: number) {
    await axios.delete(`${BASE_URL}/kb/documents/${docId}`, { headers: authHeaders() })
  },

  // --------- 定制模型（Dify 编排） ---------
  async listCustomModels() {
    const res = await axios.get<AiCustomModelItem[]>(`${BASE_URL}/custom-models`, { headers: authHeaders() })
    return res.data
  },
  async createCustomModel(payload: AiCustomModelCreate) {
    const res = await axios.post<AiCustomModelItem>(`${BASE_URL}/custom-models`, payload, { headers: authHeaders() })
    return res.data
  },
  async updateCustomModel(id: number, payload: AiCustomModelUpdate) {
    const res = await axios.put<AiCustomModelItem>(`${BASE_URL}/custom-models/${id}`, payload, { headers: authHeaders() })
    return res.data
  },
  async deleteCustomModel(id: number) {
    await axios.delete(`${BASE_URL}/custom-models/${id}`, { headers: authHeaders() })
  },

  // --------- AI 设置：功能绑定 ---------
  async getFeatureBindings(feature: AiFeatureKey) {
    const res = await axios.get<AiFeatureBindingOut>(`${BASE_URL}/feature-bindings/${feature}`, { headers: authHeaders() })
    return res.data
  },
  async saveFeatureBindings(feature: AiFeatureKey, customModelIds: number[]) {
    const res = await axios.put<AiFeatureBindingOut>(`${BASE_URL}/feature-bindings/${feature}`, { custom_model_ids: customModelIds }, { headers: authHeaders() })
    return res.data
  },

  // --------- AI 客服设置 ---------
  async getCustomerServiceSettings() {
    const res = await axios.get<AiCustomerServiceSettings>(`${BASE_URL}/customer-service/settings`, { headers: authHeaders() })
    return res.data
  },
  async updateCustomerServiceSettings(payload: Partial<AiCustomerServiceSettings>) {
    const res = await axios.put<AiCustomerServiceSettings>(`${BASE_URL}/customer-service/settings`, payload, { headers: authHeaders() })
    return res.data
  },

  // --------- AI 工作流：知识库 ---------
  async listWorkflowKnowledgeBases(feature?: string) {
    const res = await axios.get<AiKnowledgeBaseItem[]>(`${BASE_URL}/workflows/knowledge-bases`, {
      headers: authHeaders(),
      params: feature ? { feature } : undefined
    })
    return res.data
  },
  async createWorkflowKnowledgeBase(payload: AiKnowledgeBaseCreate) {
    const res = await axios.post<AiKnowledgeBaseItem>(`${BASE_URL}/workflows/knowledge-bases`, payload, { headers: authHeaders() })
    return res.data
  },
  async updateWorkflowKnowledgeBase(id: number, payload: Partial<AiKnowledgeBaseCreate>) {
    const res = await axios.put<AiKnowledgeBaseItem>(`${BASE_URL}/workflows/knowledge-bases/${id}`, payload, { headers: authHeaders() })
    return res.data
  },
  async deleteWorkflowKnowledgeBase(id: number) {
    await axios.delete(`${BASE_URL}/workflows/knowledge-bases/${id}`, { headers: authHeaders() })
  },
  async listWorkflowDocuments(kbId: number) {
    const res = await axios.get<AiKbDocument[]>(`${BASE_URL}/workflows/knowledge-bases/${kbId}/documents`, { headers: authHeaders() })
    return res.data
  },
  async uploadWorkflowDocument(kbId: number, title: string, file: File) {
    const form = new FormData()
    form.append('title', title)
    form.append('file', file)
    const res = await axios.post<AiKbDocument>(`${BASE_URL}/workflows/knowledge-bases/${kbId}/documents/upload`, form, {
      headers: { ...authHeaders() }
    })
    return res.data
  },
  async createWorkflowDocumentManual(kbId: number, payload: { title: string; content: string }) {
    const res = await axios.post<AiKbDocument>(`${BASE_URL}/workflows/knowledge-bases/${kbId}/documents/manual`, payload, { headers: authHeaders() })
    return res.data
  },
  async deleteWorkflowDocument(docId: number) {
    await axios.delete(`${BASE_URL}/workflows/documents/${docId}`, { headers: authHeaders() })
  },

  // --------- AI 工作流：应用 ---------
  async listWorkflowApps() {
    const res = await axios.get<AiWorkflowAppItem[]>(`${BASE_URL}/workflows/apps`, { headers: authHeaders() })
    return res.data
  },
  async createWorkflowApp(payload: AiWorkflowAppUpdate & { code: string; type: string; name: string }) {
    const res = await axios.post<AiWorkflowAppItem>(`${BASE_URL}/workflows/apps`, payload, { headers: authHeaders() })
    return res.data
  },
  async updateWorkflowApp(code: string, payload: AiWorkflowAppUpdate) {
    const res = await axios.put<AiWorkflowAppItem>(`${BASE_URL}/workflows/apps/${code}`, payload, { headers: authHeaders() })
    return res.data
  },
  async deleteWorkflowApp(code: string) {
    await axios.delete(`${BASE_URL}/workflows/apps/${code}`, { headers: authHeaders() })
  }
}
